File: src/pkg_py_example/pkg_py_example/simple_publisher.py
import rclpy                     # 用于与ROS 2核心进行交互的Python客户端库
from rclpy.node import Node      # 节点父类
from std_msgs.msg import String  # 发布标准的String数据  


class SimplePublisher(Node):

    def __init__(self):  # 构造函数，用于对象初始化
        super().__init__("simple_publisher")                        # 定义节点名称
        self.pub_ = self.create_publisher(String, "chatter", 10)    # 创建发布者，指定了消息类型、话题名称和缓冲区队列大小
        self.counter_ = 0                     # 用于计算发布的消息数量
        self.frequency_ = 1.0                 # 定义在主题中发布消息的频率，初始化为1Hz
        self.get_logger().info("Publishing at %d Hz" % self.frequency_)  # 打印一条info级别的log信息
        
        self.timer_ = self.create_timer(self.frequency_, self.timerCallback)  # 创建一个新的计时器对象，该计时器以指定的频率重复执行一个特定函数

    def timerCallback(self):  # 计时器回调函数，目标是每次执行时在chatter话题中发布一条新消息
        msg = String()              # 分配一个字符串变量
        msg.data = "Hello ROS 2 - counter: %d" % self.counter_  # 构造要发布的字符串
        self.pub_.publish(msg)      # 发布消息
        self.counter_ += 1          # 发布消息后计数器加1


def main():          # 主函数
    rclpy.init()     # 使用rclpy库初始化ROS2接口，并使用它的init函数

    simple_publisher = SimplePublisher()  # 创建一个SimplePublisher类的新对象
    rclpy.spin(simple_publisher)          # 使用rclpy库中定义的spin函数，保持这个节点运行，从而保持定时器和发布者持续活动
    
    simple_publisher.destroy_node()  # 当在终端中按Ctrl+C终止这个节点的执行时，还需要确保simple_publisher节点被正确销毁
    rclpy.shutdown()  # 关闭ROS2接口


if __name__ == '__main__':  # 可执行程序的入口函数
    main()  # 调用主函数

File: src/pkg_cpp_example/src/simple_publisher.cpp
#include <rclcpp/rclcpp.hpp>  // 引入ROS2客户端库的核心头文件
#include <std_msgs/msg/string.hpp>  // 引入std_msgs中String消息类型的头文件

#include <chrono>  // 引入chrono库，用于处理时间相关的功能

using namespace std::chrono_literals;  // 允许使用后缀"s"来表示秒

// 定义一个SimplePublisher类，继承自rclcpp::Node
class SimplePublisher : public rclcpp::Node
{
public:
  SimplePublisher() : Node("simple_publisher"), counter_(0)  // 构造函数，初始化节点名称和计数器
  {
    // 创建一个发布者，发布std_msgs::msg::String类型的消息到"chatter"话题，消息队列长度为10
    pub_ = create_publisher<std_msgs::msg::String>("chatter", 10);
    // 创建一个定时器，每秒调用一次timerCallback函数
    timer_ = create_wall_timer(1s, std::bind(&SimplePublisher::timerCallback, this));
    // 输出日志信息，表示发布频率为1Hz
    RCLCPP_INFO(get_logger(), "Publishing at 1 Hz");
  }

  // 定时器回调函数，当定时器触发时调用
  void timerCallback()
  {
    auto message = std_msgs::msg::String();  // 创建一个String类型的消息
    // 设置消息内容为"Hello ROS 2 - counter:"加上计数器的值，并将计数器递增
    message.data = "Hello ROS 2 - counter:" + std::to_string(counter_++);
    // 发布消息
    pub_->publish(message);
  }

private:
  // 创建一个std_msgs::msg::String类型的发布者共享指针
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
  // 创建一个定时器共享指针
  rclcpp::TimerBase::SharedPtr timer_;
  // 创建一个无符号整型计数器
  unsigned int counter_;
};

// main函数，程序的入口点
int main(int argc, char* argv[])
{
  // 初始化ROS2客户端库
  rclcpp::init(argc, argv);
  // 创建SimplePublisher节点的一个共享指针实例
  auto node = std::make_shared<SimplePublisher>();
  // 进入事件循环，等待回调函数执行
  rclcpp::spin(node);
  // 关闭ROS2客户端库
  rclcpp::shutdown();
  // 返回0，表示程序正常退出
  return 0;
}